bin/compile

```diff
@@ -1,28 +1,95 @@
 #!/bin/bash
 
+set -e
+
 BUILD_DIR=$1
-#CACHE_DIR=$2
+CACHE_DIR=$2
 ENV_DIR=$3
 BUILDPACK_DIR="$(dirname "$(dirname "$0")")"
 
+function main() {
+  export_env_dir "$ENV_DIR"
+
+  # Get latest stable version from the tailscale pkgs page
+  TAILSCALE_VERSION=${TAILSCALE_VERSION:-$(curl --silent https://pkgs.tailscale.com/stable/ | \
+    awk '/option/ {gsub("<[^>]+>", ""); print $1}' | \
+    grep -v latest | sort -r | head -1)}
+
+  # if we have a cached copy of the tailscale executable...
+  if [ -f "$CACHE_DIR/tailscale" ]; then
+    cached_version=$("$CACHE_DIR/tailscale" --version | head -1)
+
+    # ...and the cached version matches the requested version.
+    if [ "$cached_version" = "$TAILSCALE_VERSION" ]; then
+      echo "Installing Tailscale v$TAILSCALE_VERSION from cache" | major
+      copy_binaries_from_cache
+    fi
+  fi
+
+  if [ ! -f "$BUILD_DIR/bin/tailscale" ]; then
+    echo "Installing Tailscale v$TAILSCALE_VERSION" | major
+    install_from_apt
+    copy_binaries_to_cache
+    copy_binaries_from_cache
+  fi
+
+  copy_start_script
+  echo "Tailscale installed" | major
+  echo
+}
+
 function indent() {
   sed -u 's/^/       /'
 }
 
-if [[ ! -e "$ENV_DIR/TAILSCALE_AUTHKEY" ]]; then
-  echo "error: must set TAILSCALE_AUTHKEY variable in order to move forward"
-  exit 1
-fi
+function major() {
+  sed -u 's/^/-----> /'
+}
 
-# load stack-appropriate versions of the tailscale binaries into /app/bin
-echo "Installing tailscale binaries" | indent
-mkdir -p "$BUILD_DIR"/bin
-tar -xzf "$BUILDPACK_DIR/$STACK.tgz" -C "$BUILD_DIR"/bin
+export_env_dir() {
+  env_dir=$1
+  acceptlist_regex=${2:-''}
+  denylist_regex=${3:-'^(PATH|GIT_DIR|CPATH|CPPATH|LD_PRELOAD|LIBRARY_PATH)$'}
+  if [ -d "$env_dir" ]; then
+    # shellcheck disable=SC2045
```

Comment:
I love that Heroku's example code uses "bad practices" for globbing.

Reply:
heh. yeah. though, I always sort of look askance at shellcheck stuff. Sometimes I think it's just overly aggressive about various things. But yeah, I thought that was funny too. 😄